python_modules/dagster/dagster/_config/pythonic_config/type_check_utils.py

```diff
@@ -13,3 +21,12 @@ def safe_is_subclass(cls: Any, possible_parent_cls: Type) -> bool:
         # even though the isinstance check will succeed (as will inspect.isclass), for example
         # list[dict[str, str]] will raise a TypeError
         return False
+
+
+def is_optional(annotation: Type) -> bool:
+    """Returns true if the annotation is Optional[T] or Union[T, None]."""
+    return (
+        get_origin(annotation) in (Union, UnionType)
+        and len(get_args(annotation)) == 2
+        and type(None) in get_args(annotation)
+    )
```

Comment:
This comment isn't quite accurate. This returns true if _any_ of the generic arguments to `Union` is `None`, I believe.

I think this is worth:

1) Separating the cases of `Union` and `UnionType` more clearly. Without really deep knowledge of this corner of the Python type system internals it is hard to understand how your description maps to the code. 
2) Explaining to how `Optional` maps to `Union`, linking to docs if necessary.

This is the type of code that you can easily see causing bugs across different Python versions, so I think we should bias towards being super explicit and over-explaining. (That's kind of true for a bunch of the stuff in this PR).

Reply:
Fixed docstring + expanded in comments on some of the typing nuances.